### PR TITLE
[Android] TextBlocks should not receive focus when they are the only item for selectAction

### LINKED ISSUE #7864: {Linked:Bug2611817}[Accessibility] Talkback focus moves two times on the element which has select action
### Target Platforms

Android

### SDK Version

2.8.2

### Application Name

Microsoft Teams

### Problem Description

**Test Environment:**
Device: Samsung Galaxy S20
Android Version: 12
App name: Teams Production Build
App version: 1416/1.0.0.2022363703/0824

**Repro Steps:**

- Launch MS teams and login with valid credentials.
- Activate talkback.
- Navigate to 'Cortana Voice Activation' button and activate it.
- Cortini popup dialog will be open.
- Try to Give Command "Join my next meeting".
- Navigate to the 'Change meeting settings' button using swipe gesture.
- Observe talkback focuses on lands on two times on the control or not.

**Actual Results:** 
After talkback focus lands on the 'Change meeting settings' button and after next swipe focus moves to "change meeting settings" again but this time talkback does not announce the button role information.

**Expected Result:**
Talkback focus should not move two times on the "Change meeting settings" button present in "Ready to join meeting" page. As after using swipe one time focus moves to 'Change meeting settings' button only and after another swipe focus moves to the next control.

**Note: this issues can be reproduced on AC sample app with a very simple json payload**

MAS Reference: 
[MAS 1.3.2 – Meaningful Sequence](https://aka.ms/MAS1.3.2)

### Screenshots

First focus, announce "Change Meeting Settings, button"
image

Second focus, announce "Change Meeting Settings"
image

### Card JSON

```JSON
{
    "type": "AdaptiveCard",
    "version": "1.0",
    "speak": "Test Meeting title",
    "body": [
        {
            "type": "Container",
            "items": [
                {
                    "type": "ColumnSet",
                    "columns": [
                        {
                            "type": "Column",
                            "width": "stretch",
                            "items": [
                                {
                                    "type": "TextBlock",
                                    "size": "large",
                                    "text": "Test Meeting title",
                                    "wrap": true,
                                    "maxLines": 2,
                                    "spacing": "medium"
                                },
                                {
                                    "type": "TextBlock",
                                    "isSubtle": true,
                                    "text": "Tomorrow,  12:05 AM - 1:00 AM",
                                    "spacing": "none"
                                }
                            ]
                        }
                    ]
                }
            ]
        },
        {
            "type": "Container",
            "items": [
                {
                    "type": "TextBlock",
                    "size": "medium",
                    "isSubtle": true,
                    "text": "Meeting settings",
                    "spacing": "none"
                },
                {
                    "type": "ColumnSet",
                    "columns": [
                        {
                            "type": "Column",
                            "width": "auto",
                            "items": [
                                {
                                    "type": "Image",
                                    "url": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAADAAAAAwCAYAAABXAvmHAAAACXBIWXMAACxLAAAsSwGlPZapAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAJrSURBVHgB1ZnhcdowGIZfqVyvP+kGjJBOAF2jTYI7QicoTFA6QdWE3PVfmglCNmAEdwP/aH7FWP0+J5dziGwkbEvykwsHssQ9r/3ZRpbAwEiSZIIiv6a3E2i9FBgQj/IPt4CYPDVlEgPBIE/obBBHoEY+hdx9jD5Ak7xSv9OoAxyS50/i9aDTGXZ6DiFn1HmCbtmiKO4wertSSqVNHW3kGVEZMMaOLk8CM/hhoS7WS9MGW3lG1A/oHwGd/Ly4+lVts5FPzk+/0ZF8z0dyVG7PHxaQ1QE+0Fst72+qLYfkn6sEVCWS7gBFPhVlzRe4hVf0FvKepP5ksJc/oe3X+1UikesEXnGX/3L+eU5721jikv6m8MYRe35+9l1DKGocm75R0Amh4QU3eeBdZnNVHMELzvLjcpuobjPj4cecm7wo3kzr6t1EzwHca76p3k30GMBWHnxH2+zfYW3pKYCDfEt6COBPnuk4gF95psMA/uWZjgKEkWc6CBBOnmkZIKw80yJAeHnmyABxyDNHBIhHnnEMEJc84xAgPnnGMkCc8oxFgHjlmQMBAshr/ZVeM9vuDQHC7Hl1ebWCzD+UkxwLagKELZtycsNzY43Nob4UQKQvm0KUjfi738Qh1OWaJvhYNg3l53N3lW8Ks+eL+j1ND4AXTeeFRFFOohGubHSq1utNU4+m80I+Df4Rqubp8fjCpp/5vNDmFRqPJ+yyLBFHkrNPCSU/gSxWhhUaL/IZr/GWpdGSFwF6lufyTOn/BvLfqlqubaguMVkv68REwxJT/PKMGLI8I4csz/BPicqT4GHJM5IvZ3i8QmyHJs/8B3OjcjOBk8MRAAAAAElFTkSuQmCC",
                                    "horizontalAlignment": "center",
                                    "altText": "camera",
                                    "width": "12px",
                                    "height": "12px"
                                }
                            ],
                            "verticalContentAlignment": "center"
                        },
                        {
                            "type": "Column",
                            "width": "auto",
                            "items": [
                                {
                                    "type": "TextBlock",
                                    "size": "small",
                                    "isSubtle": true,
                                    "text": "Video is off",
                                    "spacing": "none"
                                }
                            ],
                            "verticalContentAlignment": "center",
                            "spacing": "small"
                        },
                        {
                            "type": "Column",
                            "width": "auto",
                            "items": [
                                {
                                    "type": "Image",
                                    "url": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAADAAAAAwCAYAAABXAvmHAAAACXBIWXMAACxLAAAsSwGlPZapAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAPDSURBVHgB7Zm9UttAFIXPSoY25gnivEBwmS72EwS6DEMS0SUVblMBTxDcpUOAJ0MX6NIBT4CcpEiH3gBnJswwWNbmruQfCWullbRykpl8DWht756jvXv3j0ETlmXV4Xvb4GiBoUlF9fFHDnz/ErXlfdu2XWiGQQPW680d+tPBTHQynNswl/Z0GillgN56A/7wM1XTVP8Vd2EstXWZMFCG3OIFTJg+D8xroLCBMGzyip/AGhh5B9BAoRAKQ8e7Rll8tO1e7yKxjVcbFiWEZtbgL9YDo+EadMCwk1Qc9C4zDmAY2/SiroIMJ6FgCLEXqIhxRtuNFJH4+4bs+zUUIczzJeEOzNv1aEmCeEHXtj85slqKDuI6SkHijVtKpaeDSYlE/KF91Ouk1VQujRYil3grq7YFG9ArXrBAA/rFCxZkoBrxggUYqE68oGID1YoXVGigevGCigwoixfrIRslqMBADvEa0GxgseIFGg0sXrxAupgLl7AjWjabF9nbP33i87Wb1gP+8Ao+PwjW4xsbKatPzW9eud2QRAOWtdkKtn0hdZimpCK94tXbnVFiDGgSX/MHKIHEgOfGn/34+p/jVN+bX3ZRArUdGWON6KN93FPZSSnAXRqosx7wKGyir5TTGVIGiT1g2ycu/ZlVzLEqq6DUgGW4QFxNK/asEF7yMcDhRBpqJp0MlM7zbLT3oOR5RICbtheekDKI+Vm0smiHCDRMUnvjng7rC0/qZlnnYe9IkBswl+zwn0pmWLEC3Y2VeMPYM+PmKRQwZR84jnPXXH26QuK3NIvvkvi30YJx/v8wK6HwOeq9U6nMTPvQ6X/94jg/7qYNSdbzMLx1cLaSelbKKSQ4tiiDfURMvDimHIlD4ukYo/POjtP/1ocCymejKpsRy3rZgFdrUmA2KQU+pjj4SVF6jXvvzD6ZxXtc/PA8MvsGRslkG4ooGZCIvyTxLRQkUXxwdzBqRwd3FpkGUmO+wI3L9Cpq/kZnAMNvq6TOKKkGlAdsYIQd0szjxGZWRETDa8Lz1+jE+Q3mjyYLiRewnOIPKUYf0a/WIK/SnS4BmBAqBidvyL+fP2xirSUVZg3Y4PKBsZ14/OZG9FQXxq/9aJrOy5yBPEcfBY1oET4hycAN4jGaeW4T7JxqBq1j/CeUNsV6JjIfUIiIOYAxui9GX3alVJQkA1cRAbkPnYJZ1cf5tIAzmryObVTE/H5AzKq+0RETEInv4i9nzsA4G3Twj/AHbmj08t/APPEDAcbYDSrEhGYc5/sg2EcAz8KV5fF7VMhvu5soIvGe5TsAAAAASUVORK5CYII=",
                                    "horizontalAlignment": "center",
                                    "width": "12px",
                                    "height": "12px"
                                }
                            ],
                            "verticalContentAlignment": "center",
                            "spacing": "large"
                        },
                        {
                            "type": "Column",
                            "width": "auto",
                            "items": [
                                {
                                    "type": "TextBlock",
                                    "size": "small",
                                    "isSubtle": true,
                                    "text": "Mic is off",
                                    "spacing": "none"
                                }
                            ],
                            "verticalContentAlignment": "center",
                            "spacing": "small"
                        }
                    ],
                    "spacing": "none"
                },
                {
                    "type": "Container",
                    "items": [
                        {
                            "type": "TextBlock",
                            "size": "small",
                            "color": "accent",
                            "text": "Change meeting settings"
                        }
                    ],
                    "selectAction": {
                        "type": "Action.Submit",
                        "data": {
                            "intent:confirm": true,
                            "change_settings": "true"
                        },
                        "title": "Change meeting settings"
                    }
                }
            ],
            "spacing": "extraLarge"
        },
        {
            "type": "Container",
            "items": [],
            "spacing": "medium"
        }
    ],
    "actions": [
        {
            "type": "Action.Submit",
            "data": {
                "intent:confirm": true
            },
            "title": "Join"
        },
        {
            "type": "Action.Submit",
            "data": {
                "intent:cancel": true
            },
            "title": "Cancel"
        }
    ]
}
```

### Sample Code Language

_No response_

### Sample Code

_No response_

### PR DESCRIPTION
**Pull request by anna-dingler**
# Related Issue

Fixes #7864

# Description

`TextBlock` should not receive focus when it is the only element within `selectAction`.

When rendering `selectAction`, if there is only one item in the `ViewGroup` and it is a TextBlock, we disable the accessibility properties.

- `setFocus(false)` handles keyboard
- `setScreenReaderFocusable(false)` and `setImportantForAccessibility(View.IMPORTANT_FOR_ACCESSIBILITY_NO)` handles swipe gesture.

# Sample Card

```JSON
{
    "type": "AdaptiveCard",
    "version": "1.0",
    "speak": "Test Meeting title",
    "body": [
        {
            "type": "Container",
            "items": [
                {
                    "type": "TextBlock",
                    "size": "small",
                    "color": "accent",
                    "text": "Change meeting settings"
                }
            ],
            "selectAction": {
                "type": "Action.Submit",
                "data": {
                    "intent:confirm": true,
                    "change_settings": "true"
                },
                "title": "Change meeting settings"
            }
        }
    ]
}
```

# How Verified

Verified on the Android Visualizer app.

(Tab key is pressed twice in this gif)
androidTalkBack

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/AdaptiveCards/pull/8025)

### CODE REVIEW

**Review by paulcam206 — COMMENTED** [review]
code looks good -- @almedina-ms, is this a reasonable pattern?

**Comment by longlong-cs** [comment]
@almedina-ms could you help review this pull request?

File: source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/readonly/TextBlockRenderer.java [hunk 1]
```diff
@@ -225,32 +225,26 @@ private void applyAccessibilityProperties(TextView textView, Context context, Re
         }
         textView.setImportantForAccessibility(View.IMPORTANT_FOR_ACCESSIBILITY_YES);
 
-        if (textHandleResult.getHasLinks())
-        {
+        if (textHandleResult.getHasLinks()) {
```
Comment (jwoo-msft) [inline source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/readonly/TextBlockRenderer.java:228]: nit: curly bracket placement has changed; and different from the rest of codes.

**Review by jwoo-msft — APPROVED** [review]
regular text won't be focusable? will it cause voice over not to announce a11y labels of regular text without links?

**Comment by anna-dingler** [comment]
> regular text won't be focusable? will it cause voice over not to announce a11y labels of regular text without links?

@almedina-ms Do you have any thoughts here?

Regular text should not be focusable in the general case. Tab order only includes elements that have an interaction.
That being said, I'm not sure what the impact would be on TalkBack and why this was added originally.

**Review by almedina-ms — APPROVED** [review]

File: source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/readonly/TextBlockRenderer.java [hunk 2]
```diff
@@ -225,32 +225,26 @@ private void applyAccessibilityProperties(TextView textView, Context context, Re
...
+                    public void onAccessibilityStateChanged(boolean b) {
+                        boolean isEnabled = am.isEnabled();
+                        if (b && isEnabled) {
+                            textView.setFocusable(true);
+                        } else {
+                            textView.setFocusable(false);
+                        }
+                    }
+                });
+                textView.setFocusable(am.isEnabled());
```
Comment (almedina-ms) [inline source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/readonly/TextBlockRenderer.java:245]: can you verify that, by moving this line here, cases such as 

"this is a string [link](url) and this is another [link](url)"

also behaves as expected?

Comment (anna-dingler) [reply source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/readonly/TextBlockRenderer.java:245]: Yes, I'll check on this

**Comment by longlong-cs** [comment]
Hi @anna-dingler I pulled your branch and build locally, I can still reproduce these two issues #7864, #7959, so I'm wondering whether this PR can fix them.

File: source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/readonly/TextBlockRenderer.java (hunk 2 above)
Comment (anna-dingler) [reply source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/readonly/TextBlockRenderer.java:245]: Hey @almedina-ms, I tested this scenario, and it does not work as expected.

I can move the focus logic to be within the `hasLinks` block; however, the individual spans are not accessible via tab key and pressing enter does not launch either url. 

I am seeing this issue on main as well, so I will file a separate bug for this. Seems like the problem here is that `URLSpan` is not a view, so we cannot use the `SetFocus` method.

**Comment by anna-dingler** [comment]
> Hi @anna-dingler I pulled your branch and build locally, I can still reproduce these two issues #7864, #7959, so I'm wondering whether this PR can fix them.

Hi @longlong-cs,

Can you provide any information about how you are testing the change locally (test app, etc.)?

I am attaching gifs here of what I am seeing on my machine running code from the branch `anna/androidDoubleFocus` and using the sample payloads provided in the bugs.

### Issue 7864 - TextBlock receiving double focus with selectAction

Once we are focused on the "Options" button:
1. Pressing tab moves focus to "Change meeting settings" link
2. Pressing tab again moves focus to "Join" button

doubleFocus

### Issue 7864 - TextBlocks without links receiving focus

Once we are focused on the "Options" button:
1. Pressing tab moves focus to the "Last, First" EditText

None of the prior TextBlocks receive focus.

noTextBlockFocus

**Comment by anna-dingler** [comment]
> regular text won't be focusable? will it cause voice over not to announce a11y labels of regular text without links?

I reverted back to the original approach to fix this :)

File: source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/readonly/ContainerRenderer.java [hunk 3]
```diff
@@ -308,6 +309,25 @@ public static void setSelectAction(RenderedAdaptiveCard renderedCard, BaseAction
             view.setOnClickListener(new BaseActionElementRenderer.SelectActionOnClickListener(renderedCard, selectAction, cardActionHandler));
 
             applyTitleAndTooltip(selectAction, view);
+
+            if (view instanceof ViewGroup)
```
Comment (anna-dingler) [inline source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/readonly/ContainerRenderer.java:313]: @jwoo-msft I just made the generalization changes here

File: source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/readonly/TextBlockRenderer.java (hunk 2 above)
Comment (anna-dingler) [reply source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/readonly/TextBlockRenderer.java]: Created issue #8077

File: source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/readonly/ContainerRenderer.java [hunk 4]
```diff
@@ -308,6 +309,25 @@ public static void setSelectAction(RenderedAdaptiveCard renderedCard, BaseAction
...
+            {
+                ViewGroup group = (ViewGroup) view;
+                if (group.getChildCount() == 1)
+                {
+                    View childView = group.getChildAt(0);
+                    if (childView.isFocusable())
+                    {
+                        childView.setFocusable(false);
+
+                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P)
```
Comment (KesemZ) [inline source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/readonly/ContainerRenderer.java:323]: I suggest adding a comment as it is not clear how the build version is related here.

Comment (anna-dingler) [reply source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/readonly/ContainerRenderer.java:325]: Just added a comment explaining this!

**Review by KesemZ — APPROVED** [review]

**Pull request merged**